homeassistant/components/sunweg/__init__.py

```diff
@@ -0,0 +1,193 @@
+    """Unload a config entry."""
+    hass.data[DOMAIN].pop(entry.entry_id)
+    if len(hass.data[DOMAIN]) == 0:
+        hass.data.pop(DOMAIN)
+    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+class SunWEGData:
+    """The class for handling data retrieval."""
+
+    def __init__(
+        self,
+        api: APIHelper,
+        plant_id: int,
+    ) -> None:
+        """Initialize the probe."""
+
+        self.api = api
+        self.plant_id = plant_id
+        self.data: Plant = None
+        self.previous_values: dict = {}
+
+    @Throttle(SCAN_INTERVAL)
+    def update(self) -> None:
+        """Update probe data."""
+        _LOGGER.debug("Updating data for plant %s", self.plant_id)
+        try:
+            self.data = self.api.plant(self.plant_id)
+            for inverter in self.data.inverters:
+                self.api.complete_inverter(inverter)
+        except json.decoder.JSONDecodeError:
+            _LOGGER.error("Unable to fetch data from SunWEG server")
+        _LOGGER.debug("Finished updating data for plant %s", self.plant_id)
+
+    def get_api_value(
+        self,
+        variable: str,
+        device_type: str,
+        inverter_id: int = 0,
+        deep_name: str | None = None,
+    ):
+        """Retrieve from a Plant the desired variable value."""
+        if device_type == "total":
+            return self.data.__dict__.get(variable)
+
+        inverter_list = [i for i in self.data.inverters if i.id == inverter_id]
+        if len(inverter_list) == 0:
+            return None
+        inverter = inverter_list[0]
+
+        if device_type == "inverter":
+            return inverter.__dict__.get(variable)
+        if device_type == "phase":
+            for phase in inverter.phases:
+                if phase.name == deep_name:
+                    return phase.__dict__.get(variable)
+        elif device_type == "string":
+            for mppt in inverter.mppts:
+                for string in mppt.strings:
+                    if string.name == deep_name:
+                        return string.__dict__.get(variable)
+        return None
+
+    def get_data(
+        self,
+        entity_description: SunWEGSensorEntityDescription,
+        device_type: str,
+        inverter_id: int = 0,
+        deep_name: str | None = None,
+    ) -> StateType | datetime.datetime:
+        """Get the data."""
+        _LOGGER.debug(
+            "Data request for: %s",
+            entity_description.name,
+        )
+        variable = entity_description.api_variable_key
+        previous_metric = entity_description.native_unit_of_measurement
+        api_value = self.get_api_value(variable, device_type, inverter_id, deep_name)
+        previous_value = self.previous_values.get(variable)
+        return_value = api_value
+        if entity_description.api_variable_metric is not None:
+            entity_description.native_unit_of_measurement = self.get_api_value(
```

Comment:
Entity descriptions should not be modified. They are considered frozen.

Generally if the attribute needs to be calculated dynamically it doesn't belong in the entity description. Then it's better to set it inside the entity or pass it to the entity after calculating it.